client/emv/emvcore.c

```diff
@@ -23,11 +94,370 @@ void TLVPrintFromBuffer(uint8_t *data, int datalen) {
+		return res;
+	}
+	
+	if (*ResultLen < 2) {
+		PrintAndLog("SELECT ERROR: returned %d bytes", *ResultLen);
+		return 5;
+	}
+	
+	*ResultLen -= 2;
+	isw = Result[*ResultLen] * 0x0100 + Result[*ResultLen + 1];
+	if (sw)
+		*sw = isw;
+
+	if (isw != 0x9000) {
+		if (APDULogging)
+			PrintAndLog("SELECT ERROR: [%4X] %s", isw, GetAPDUCodeDescription(*sw >> 8, *sw & 0xff));
+		return 5;
+	}
+
+	// add to tlv tree
+	if (tlv) {
+		struct tlvdb *t = tlvdb_parse_multi(Result, *ResultLen);
+		tlvdb_add(tlv, t);
+	}
+	
+	return 0;
+}
+
+int EMVSelectPSE(bool ActivateField, bool LeaveFieldON, uint8_t PSENum, uint8_t *Result, size_t MaxResultLen, size_t *ResultLen, uint16_t *sw) {
+	uint8_t buf[50] = {0};
+	*ResultLen = 0;
+	int len = 0;
+	int res = 0;
+	switch (PSENum) {
+		case 1:
+			param_gethex_to_eol(PSElist[1], 0, buf, sizeof(buf), &len);
+			break;
+		case 2:
+			param_gethex_to_eol(PSElist[0], 0, buf, sizeof(buf), &len);
+			break;
+		default:
+			return -1;
+	}
+	
+	// select
+	res = EMVSelect(ActivateField, LeaveFieldON, buf, len, Result, MaxResultLen, ResultLen, sw, NULL);
+
+	return res;
+}
+
+int EMVSearchPSE(bool ActivateField, bool LeaveFieldON, bool decodeTLV, struct tlvdb *tlv) {
+	uint8_t data[260] = {0};
+	size_t datalen = 0;
+	uint16_t sw = 0;
+	int res;
+
+	// select PPSE
+	res = EMVSelectPSE(ActivateField, true, 2, data, sizeof(data), &datalen, &sw);
+
+	if (!res){
+		struct tlvdb *t = NULL;
+		t = tlvdb_parse_multi(data, datalen);
+		if (t) {
+			int retrycnt = 0;
+			struct tlvdb *ttmp = tlvdb_find_path(t, (tlv_tag_t[]){0x6f, 0xa5, 0xbf0c, 0x61, 0x00});
+			if (!ttmp)
+				PrintAndLog("PPSE don't have records.");
+			
+			while (ttmp) {
+				const struct tlv *tgAID = tlvdb_get_inchild(ttmp, 0x4f, NULL);
+				if (tgAID) {
+					res = EMVSelect(false, true, (uint8_t *)tgAID->value, tgAID->len, data, sizeof(data), &datalen, &sw, tlv);
+
+					// retry if error and not returned sw error
+					if (res && res != 5) {
+						if (++retrycnt < 3){
+							continue;
+						} else {
+							// card select error, proxmark error
+							if (res == 1) {
+								PrintAndLog("Exit...");
+								return 1;
+							}
+							
+							retrycnt = 0;
+							PrintAndLog("Retry failed [%s]. Skiped...", sprint_hex_inrow(tgAID->value, tgAID->len));
+						}
+						
+						// next element
+						ttmp = tlvdb_find_next(ttmp, 0x61);
+						continue;
+					}
+					retrycnt = 0;
+
+					// all is ok
+					if (decodeTLV){
+						PrintAndLog("%s:", sprint_hex_inrow(tgAID->value, tgAID->len));
+						TLVPrintFromBuffer(data, datalen);
+					}
+				}
+				
+				ttmp = tlvdb_find_next(ttmp, 0x61);
+			}
+
+			tlvdb_free(t);
+		} else {
+			PrintAndLog("PPSE ERROR: Can't get TLV from response.");
+		}		
+	} else {
+		PrintAndLog("PPSE ERROR: Can't select PPSE AID. Error: %d", res);
+	}
+	
+	if(!LeaveFieldON)
+		DropField();
+	
+	return res;
+}
+
+int EMVSearch(bool ActivateField, bool LeaveFieldON, bool decodeTLV, struct tlvdb *tlv) {
+	uint8_t aidbuf[50] = {0};
```

Comment:
...

Reply:
I can unify this but it enoughfixed. 50 bytes = 100hex symbols. it enough for AID (ususally they have only 7-10 bytes)